designs/2019-eslintrc-improvements/major-02-plugin-resolution-change.md

```diff
@@ -0,0 +1,156 @@
+
+<table><td>
+💡 <b>Example</b>:
+<pre lang="js">
+// eslint-config-foo
+module.exports = {
+    plugins: ["a-plugin"],
+    parser: "./lib/parser",
+    env: {
+        "a-plugin/env": true
+    },
+    rules: {
+        eqeqeq: "error",
+        "a-plugin/x": "error",
+        "a-plugin/y": "error"
+    }
+}
+</pre>
+<code>config.withConvert("eslint-config-foo", { mapPluginName: id => `foo::${id}` })</code> method with the above config returns as:
+<pre lang="jsonc">
+[
+    {
+        "plugins": {
+            // Renaming with absolute path.
+            "foo::a-plugin": "/path/to/node_modules/eslint-config-foo/node_modules/eslint-plugin-a-plugin/index.js"
+        },
+        // absolute path.
+        "parser": "/path/to/node_modules/eslint-plugin-foo/lib/parser.js",
+        "env": {
+            "foo::a-plugin/env": true
+        },
+        "rules": {
+            "eqeqeq": "error",
+            "foo::a-plugin/x": "error",
+            "foo::a-plugin/y": "error"
+        }
+    }
+]
+</pre>
+</td></table>
+
+The `config.withConvert(request, options)` method loads `extends` property and flattens `extends` property and `overrides` property recursively. Then it converts all plugin names in the configs.
+
+> TODO: AND expression in configs for `files`/`excludedFiles`. Maybe `test: Array<{files:string[],excludedFiles?:string[]}>`?
+
+## Documentation
+
+This change needs migration guide because of a breaking change.
+
+- If a shareable config and your package, or two shareable configs, depend on the incompatible versions of the same plugin, a package manager fails to dedupe the plugin, then ESLint throws "plugin conflict" error. If you encouhtered the "plugin conflict" error, please try to solve the incompatible version problem. If you cannot solve that, use `@eslint/config` package to tweak your configuration.
+
+## Drawbacks
+
+- "conflict error" may cause new confusion.
+- It can be broken without major version of shareable configs because of lacking [Robustness Guarantee].
+
+## Backwards Compatibility Analysis
+
+Because this is after [#7], the effect of the plugin resolution change is limited.
+
+If people are using multiple shareable configs which depend on plugins via `dependencies` field of `package.json`, a package manager may fail to dedupe the plugins. However, most shareable configs are using `peerDependencies` to depend on plugins currently.
+
+## Alternatives
+
+- [#9] is the alternative. But double duplicate features cause confusion for the ecosystem. For newcomers, a mix of articles about two config systems makes hard to understand ESLint. For non-English users, the official document is far.
+- [#5] and [#14] are alternatives about the plugin resolution change. Those are more strict, but those need some complex logic.
+
+## Open Questions
+
+- <a href="#q-robustness-guarantee" id="robustness-guarantee">❓</a> Should we keep [Robustness Guarantee] for the plugin conflict error?<br>
```

Comment:
Could we make a consensus to go with the current naive strategy ([lib/lookup/config-array.js#L76-L88](https://github.com/eslint/eslint/blob/153640180a8944af3a1c488462ed30d0c215f5ed/lib/_lookup/config-array.js#L76-L88)) for now? I'm suspecting if that is rare case enough or not. Maybe we can confirm that with alpha versions?

Reply:
No, I don't think so.

* I think this is the type of conflict that would take some time to appear (e.g. it would only happen after shareable configs start listing plugins as `dependencies` rather than `peerDependencies`, and even then only after a specific combination of events). So I don't think we would catch this problem in alpha versions because many shareable configs wouldn't have migrated yet.
* If we start allowing this, it will be very difficult to change later because a lot of users will probably be relying on it. So we would end up stuck in a place where ESLint randomly breaks sometimes.

Reply:
> * I think this is the type of conflict that would take some time to appear (e.g. it would only happen after shareable configs start listing plugins as `dependencies` rather than `peerDependencies`, and even then only after a specific combination of events). So I don't think we would catch this problem in alpha versions because many shareable configs wouldn't have migrated yet.

I expected it to happens if listing plugins as `dependencies` rather than `peerDependencies` AND version ranges of the plugins are conflicted. We should raise an error in such a case rather. Even if a package manager failed to dedupe by a bug, clearing cache solves the problem.

> * If we start allowing this, it will be very difficult to change later because a lot of users will probably be relying on it. So we would end up stuck in a place where ESLint randomly breaks sometimes.

How about:

1. publish this with naive conflict handling and deprecate cascading configs together. The [`extends` in `overrides`](https://github.com/eslint/rfcs/blob/eslintrc-improvements/designs/2019-eslintrc-improvements/minor-02-extends-in-overrides.md) enhancement should help the migration to the glob-based configration.
2. remove cascading configs and implement strict conflict handling together. Removing cascading configs makes implementing it easily.

Reply:
> I expected it to happens if listing plugins as `dependencies` rather than `peerDependencies` AND version ranges of the plugins are conflicted. We should raise an error in such a case rather.

I agree that this is usually when a conflict would happen, but the problem is that this could start happening without any changes by the end user, causing their config to suddenly break. Naive conflict handling would effectively work by raising an error when things are already broken, rather than preventing things from breaking in the first place. Lots of people use ESLint as part of important infrastructure, and I don't think we should ship a design that's inherently fragile.

From what I can tell, we have two robust choices (assuming we want to keep the constraint that plugins names are globally unique):

* Shareable configs have control over which plugin versions get used. We impose constraints on the user's config to ensure that shareable configs won't inadvertently break users' configs when they decide to exert that control. (This is "strict conflict handling" from https://github.com/eslint/rfcs/pull/14 in various forms.)
* End users have control over which plugin versions get used. Shareable configs can't update plugins without action from the end user, which avoids conflicts but adds a burden for end users. (This is the current `peerDependencies` solution.)

The "naive conflict handling" solution would give the illusion that shareable configs have control over plugin versions, when in fact they wouldn't be able to update their plugins without the risk of breaking users. Effectively, we would be telling shareable config authors to just hope nothing goes wrong, and I don't think stable tools should do things like that.

To be honest, I don't think the current `peerDependencies` requirement is as big of a problem now that https://github.com/eslint/eslint/issues/10125 is fixed. Creating a robust `dependencies` solution seems like it might require imposing too many constraints on users to be worthwhile. Maybe we should address the ergonomics issues of `peerDependencies` with smaller solutions like `postinstall` scripts and better error messages for missing plugins.

Reply:
Okay, I will remove [Plugin Resolution Change](https://github.com/eslint/rfcs/blob/eslintrc-improvements/designs/2019-eslintrc-improvements/major-02-plugin-resolution-change.md) proposal from this RFC because now I have another solution the discussion with you has given me.

Thank you!